Add eval : Product Information Extraction

abrinkmann opened the pull request:
# Thank you for contributing an eval! ♥️

🚨 Please make sure your PR follows these guidelines, **failure to follow the guidelines below will result in the PR being closed automatically**. Note that even if the criteria are met, that does not guarantee the PR will be merged nor GPT-4 access be granted. 🚨

**PLEASE READ THIS**:

In order for a PR to be merged, it must fail on GPT-4. We are aware that right now, users do not have access, so you will not be able to tell if the eval fails or not. Please run your eval with GPT-3.5-Turbo, but keep in mind as we run the eval, if GPT-4 gets higher than 90% on the eval, we will likely reject it since GPT-4 is already capable of completing the task.

We plan to roll out a way for users submitting evals to see the eval performance on GPT-4 soon. Stay tuned! Until then, you will not be able to see the eval performance on GPT-4. **Starting April 10, the minimum eval count is 15 samples, we hope this makes it easier to create and contribute evals.**

Also, please note that we're using **Git LFS** for storing the JSON files, so please make sure that you move the JSON file to Git LFS before submitting a PR. Details on how to use Git LFS are available [here](https://git-lfs.com).

## Eval details 📑

### Eval name

product-information-extraction

### Eval description

The task is to extract the attribute value of an attribute specified by the user from a product title. The eval contains 562 samples of product titles from three different product categories.

### What makes this a useful eval?

Structured product data in the form of attribute/value pairs is the foundation of many e-commerce applications such as faceted product search, product comparison, and product recommendation. Product offers often only contain textual descriptions of the product attributes in the form of titles or free text. Hence, extracting attribute/value pairs from textual product titles is an essential enabler for e-commerce applications.

The samples (562) in this eval are sampled from the [MAVE ](https://github.com/google-research-datasets/MAVE/) dataset that has been derived from a public collection of product offers from the e-commerce platform Amazon. The samples are drawn from product offers that belong to three product categories and aim at extracting attribute values of 15 different attributes. The eval contains samples for **four different prompt designs including zeroshot, one-shot and three-shot** prompts.

The product information extraction task is usually evaluated with the metrics Precision, Recall and F1-Score on exactly matching extracted attribute values. Accuracy is not an optimal evaluation measure because the model's performance regarding in-correctly or missing attribute values is difficult to interpret through the evaluation measure accuracy.

## Criteria for a good eval ✅

Below are some of the criteria we look for in a good eval. In general, we are seeking cases where the model does not do a good job despite being capable of generating a good response (note that there are some things large language models cannot do, so those would not make good evals).

Your eval should be:

- [x] Thematically consistent: The eval should be thematically consistent. We'd like to see a number of prompts all demonstrating some particular failure mode. For example, we can create an eval on cases where the model fails to reason about the physical world.
- [x] Contains failures where a human can do the task, but either GPT-4 or GPT-3.5-Turbo could not.
- [x] Includes good signal around what is the right behavior. This means either a correct answer for `Basic` evals or the `Fact` Model-graded eval, or an exhaustive rubric for evaluating answers for the `Criteria` Model-graded eval.
- [x] **Include at least 15 high-quality examples.**

If there is anything else that makes your eval worth including, please document it below.

### Unique eval value

The data in the samples is real-world data from the e-commerce platform Amazon. It has been subsampled from a benchmark used in the product information extraction research community and thus properly represents this task. Research papers are available that describe the data set and the techniques used for prompt engineering.

## Eval structure 🏗️

Your eval should

- [x] Check that your data is in `evals/registry/data/{name}`
- [x] Check that your YAML is registered at `evals/registry/evals/{name}.yaml`
- [x] Ensure you have the right to use the data you submit via this eval

(For now, we will only be approving evals that use one of the existing eval classes. You may still write custom eval classes for your own cases, and we may consider merging them in the future.)

## Final checklist 👀

### Submission agreement

By contributing to Evals, you are agreeing to make your evaluation logic and data under the same MIT license as this repository. You must have adequate rights to upload any data used in an Eval. OpenAI reserves the right to use this data in future service improvements to our product. Contributions to OpenAI Evals will be subject to our usual Usage Policies (<https://platform.openai.com/docs/usage-policies>).

- [x] I agree that my submission will be made available under an MIT license and complies with OpenAI's usage policies.

### Email address validation

If your submission is accepted, we will be granting GPT-4 access to a limited number of contributors. Access will be given to the email address associated with the commits on the merged pull request.

- [x] I acknowledge that GPT-4 access will only be granted, if applicable, to the email address used for my merged pull request.

### Limited availability acknowledgment

We know that you might be excited to contribute to OpenAI's mission, help improve our models, and gain access to GPT-4. However, due to the requirements mentioned above and the high volume of submissions, we will not be able to accept all submissions and thus not grant everyone who opens a PR GPT-4 access. We know this is disappointing, but we hope to set the right expectation before you open this PR.

- [x] I understand that opening a PR, even if it meets the requirements above, does not guarantee the PR will be merged nor GPT-4 access be granted.

### Submit eval

- [x] I have filled out all required fields of this form
- [x] I have used **Git LFS** for the Eval JSON data
- [x] (Ignore if not submitting code) I have run `pip install pre-commit; pre-commit install` and have verified that `black`, `isort`, and `autoflake` are running when I commit and push

Failure to fill out all required fields will result in the PR being closed.

### Eval JSON data

Since we are using Git LFS, we are asking eval submitters to add in as many Eval Samples (at least 5) from their contribution here:

<details>
  <summary>View evals in JSON</summary>

  ### Eval
  ```jsonl
  {"input": [{"role": "system", "content": "Extract all attribute-value pairs from the product title. Afterwards, return only the exact attribute value of the requested attribute. Your answer must only contain the attribute value. If the attribute is not applicable or the attribute value cannot be derived from the product title, respond with 'I do not know.'."}, {"role": "user", "content": "SanDisk 64 MB CompactFlash Card \n Requested Attribute: Capacity"}], "ideal": ["64 MB", "64MB"]}
{"input": [{"role": "system", "content": "Extract all attribute-value pairs from the product title. Afterwards, return only the exact attribute value of the requested attribute. Your answer must only contain the attribute value. If the attribute is not applicable or the attribute value cannot be derived from the product title, respond with 'I do not know.'."}, {"role": "user", "content": "Lenovo IBM R51 ThinkPad Laptop ( 28892RU ) \n Requested Attribute: Weight"}], "ideal": ["I do not know."]}
{"input": [{"role": "system", "content": "Extract all attribute-value pairs from the product title. Afterwards, return only the exact attribute value of the requested attribute. Your answer must only contain the attribute value. If the attribute is not applicable or the attribute value cannot be derived from the product title, respond with 'I do not know.'."}, {"role": "user", "content": "Transcend 2 GB MMCplus Flash Memory Card TS2GMMC4 \n Requested Attribute: SD Format"}], "ideal": ["I do not know."]}
{"input": [{"role": "system", "content": "Extract all attribute-value pairs from the product title. Afterwards, return only the exact attribute value of the requested attribute. Your answer must only contain the attribute value. If the attribute is not applicable or the attribute value cannot be derived from the product title, respond with 'I do not know.'."}, {"role": "user", "content": "HP 612 2.3 MP Digital Camera with 2x Optical Zoom \n Requested Attribute: Resolution"}], "ideal": ["2.3 MP", "2.3MP"]}
{"input": [{"role": "system", "content": "Extract all attribute-value pairs from the product title. Afterwards, return only the exact attribute value of the requested attribute. Your answer must only contain the attribute value. If the attribute is not applicable or the attribute value cannot be derived from the product title, respond with 'I do not know.'."}, {"role": "user", "content": "Olympus Camedia C-3020 3MP Digital Camera w/ 3x Optical Zoom \n Requested Attribute: Optical Zoom"}], "ideal": ["3x Optical Zoom", "3x"]}

  ```
</details>

usama-openai reviewed (changes requested):
Thanks for the contribution. Both GPT-3.5 and GPT-4 are performing with 100% accuracy on all four evaluations, which means the task is too easy for the model. I would like to request some improvements.

1. All four evals have the same kind of data and are performing the same kind of task, so I would recommend you keep only one evaluation. In my opinion, you should only keep `match_product-information-extraction_generated-knowledge_one-shot`.

2. Eval name is too long. Kindly make it shorter.

3. Kindly add a little complexity to the task. Here are a few suggestions for you:
    - Consider the following sample:

        `Lenovo Z50-70 Laptop - Intel Core i7-4510U up to 3.1GHz 8GB DDR3 RAM, 500GB HDD, 15.6" FHD 1920x1080 Display, USB 3.0, Bluetooth, 802.11ac WiFi, Windows 8.1/10 Requested Attribute: Resolution`

        Instead of asking for just resolution, you can say something like, `I'm looking for a laptop with a larger than 15-inch screen and 1080p resolution. Does this product fulfill my requirements?`

    - You can provide a description of two or more products along with the user requirements and ask the model which product is more suitable according to the user's requirements.

    You can also think of some other improvements.

We would love to review the PR again after the suggested changes.

abrinkmann commented:
Thank you for the review @usama-openai.
According to your suggestions, I made the following changes:
1) The records for the four evals are now selected such that neither GPT3.5 nor GPT4 score above 90% accuracy.
2) The name of the eval is shortened.

The eval aims at evaluating LLMs on the tasks of product information extraction, which deals with extracting attribute values from product titles. Your suggested extension goes beyond product information extraction. Therefore, I suggest putting a task in which the model is asked to recommend a product that fulfils a set of requirements into another eval.

I am looking forward to your comments.

usama-openai reviewed (changes requested):
Thanks for implementing the requested changes. I would like to request a few more changes:

1. Kindly merge `zeroshot/generated_knowledge_flash_memory_cards_samples.jsonl` and `zeroshot/generated_knowledge_laptops_samples.jsonl` datasets to a single dataset and `match_product-ie_generated-knowledge_zero-shot_flash-memory-cards` and `match_product-ie_generated-knowledge_zero-shot_laptops`, evals to a single eval.
2. Kindly merge `fewshot/generated_knowledge_one_shot_flash_memory_cards_samples.jsonl` and `fewshot/generated_knowledge_one_shot_laptops_samples.jsonl` datasets to a single dataset and `match_product-ie_generated-knowledge_one-shot_flash-memory-cards` and `match_product-ie_generated-knowledge_one-shot_laptops`, evals to a single eval.
3. Use a shorter name for evals. You can use `product_information_extraction_zero_shot` and `product_information_extraction_one_shot` for eval names.
4. The samples `35 - 38` in `generated_knowledge_one_shot_flash_memory_cards_samples.jsonl` and `generated_knowledge_one_shot_laptops_samples.jsonl` files are supposed to have "`I don't know`" as an ideal answer, and you need to fix this.

abrinkmann commented:
Thanks, @usama-openai 
The eval is updated as requested. 
Additionally, I added some challenging examples and removed simple ones to make the eval more challenging.

usama-openai reviewed (approved):
This PR looks in good shape now. I'm approving this PR.
